RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc

```diff
@@ -291,12 +296,28 @@ void PixelThresholdClusterizer::copy_to_buffer(DigiIterator begin, DigiIterator
     */
 
     if (adc >= thePixelThreshold) {
-      theBuffer.set_adc(row, col, adc);
-      // VV: add pixel to the fake list. Only when running on digi collection
-      if (di->flag() != 0)
-        theFakePixels[row * theNumOfCols + col] = true;
-      if (adc >= theSeedThreshold)
-        theSeeds.push_back(SiPixelCluster::PixelPos(row, col));
+      thePixelOccurrence[theBuffer.index(row, col)]++;                     // increment the occurrence counter
+      uint8_t occurrence = thePixelOccurrence[theBuffer.index(row, col)];  // get the occurrence counter
+
+      switch (occurrence) {
+        // the 1st occurrence (standard treatment)
+        case 1:
+          theBuffer.set_adc(row, col, adc);
+          // VV: add pixel to the fake list. Only when running on digi collection
+          if (di->flag() != 0)
+            theFakePixels[row * theNumOfCols + col] = true;
+          if (adc >= theSeedThreshold)
+            theSeeds.push_back(SiPixelCluster::PixelPos(row, col));
+          break;
+
+        // the 2nd occurrence (duplicate pixel: reset the buffer to 0 and remove from the list of seed pixels)
+        case 2:
+          theBuffer.set_adc(row, col, 0);
+          std::remove(theSeeds.begin(), theSeeds.end(), SiPixelCluster::PixelPos(row, col));
+          break;
+
+          // in case a pixel appears more than twice, nothing needs to be done because it was already removed at the 2nd occurrence
```

Comment:
Shouldn't this be done also if `adc < thePixelThreshold` ?

If the duplicates are likely due to noise, they should be removed also when there is only one above threshold.

Reply:
Also, the check is wrong, because nothing is being done for the 3rd, 4th, _etc._ pixel.

Reply:
@ferencek can you please have a look to this ?

Reply:
> Shouldn't this be done also if `adc < thePixelThreshold` ?
> 
> If the duplicates are likely due to noise, they should be removed also when there is only one above threshold.

In principle, yes. However, the current pixel threshold is 10 and all pixels with the electric charge (for some historical reason, the variable name is `adc` but it actually represent charge) < 100 have their charge set to 100. This effectively means that there are no pixels with charge below the pixel threshold and hence all duplicates will be caught.

Reply:
Ah, I see:
  - at the 1st occurrence the data from the pixel is used;
  - at the 2nd occurrence, what was done at the 1st occurrence is undone, and the pixel itself is ignored;
  - at the 3rd occurrence and later, the pixel is ignored.

Reply:
>  Ah, I see:
>    * at the 1st occurrence the data from the pixel is used;
>    * at the 2nd occurrence, what was done at the 1st occurrence is undone, and the pixel itself is ignored;
>    * at the 3rd occurrence and later, the pixel is ignored.

is this the same logic as in https://github.com/cms-sw/cmssw/pull/38946 ?
I thought *all* duplicates are removed on the GPU code.

Reply:
Yes (if I understood correctly the logic) because
  - the 1st time a pixel is seen, it is added normally
  - the 2nd time the same pixel it is seen (a duplicate), the previous occurrence is removed, and the second occurrence is ignored
  - the 3rd time the same pixel is seen (and any later one), it is just ignored

The end results should be the same as with the GPU logic. The extra complexity there comes from the need to use atomic operations, and the use of a bit-field instead of a full word to make the module occupancy fit in 64 kB.

Reply:
are we guaranteed that the same "first" pixel is seen in both workflow?

Reply:
> the previous occurrence is removed

OK, I missed this. Fine then.